pkg/kv/kvserver/tenantrate/factory.go

```diff
@@ -121,6 +121,7 @@ func (rl *LimiterFactory) Release(lim Limiter) {
 	}
 	if rcLim.refCount--; rcLim.refCount == 0 {
 		l.metrics.destroy()
+		l.qp.Close("released")
```

Comment:
Does this strictly prevent the race? The other end may notice the closed channel too late, and still write to the metric, no?

I wonder if the proper solution to this would be ref-counting not only at the `Replica` level, but at the request level. I.e. make it so that `Send` gets a ref-incremented `Limiter`, and `Releases` it in the end; instead of relying on `Replica` to do so.

Reply:
Ah, I see, we rely on the fact that there is only one `Gauge` metric which also happens to be related to synchronization. We close the channel, and then some time after the waiters decrement the counter. In the meantime, we only `Unlink`. The counter eventually will stabilize.

Any generic gauge addition would break things again. Worth noting this and TODOing.

Reply:
👍🏽 Did you want to add a TODO?

Reply:
I think you covered it well in the comment next to `r.tenantLimiter`. Injected a TODO there.